### PR TITLE
Closes #1755 `derive_extreme_event()`

### LINKED ISSUE #1755: Feature Request:  Implement `derive_extreme_event()`
### Feature Idea

```r
derive_extreme_event(
  dataset,
  by_vars = NULL,
  events,
  order,
  mode,
  set_values_to
)
```
New admiral function to derive best/worst answer from `ADQS` workflow discussion https://github.com/pharmaverse/admiral/issues/1279

what do you think of the function name? @bundfussr @bms63 @DeclanHodges @nbrucken17 
it's similar to `derive_extreme_records()`, so I used `derive_extreme_answer()` instead of `derive_param_??`. also borrowed `mode` from `derive_extreme_records()`. 

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by kaz462**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/derive_extreme_event.R [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  ## Create a dataset (selected_records) from `events`
+  condition_ls <- map(events, 1)
+  set_values_to_ls <- map(events, 2)
+  selected_records_ls <- map2(
+    condition_ls, set_values_to_ls,
+    function(x, y) {
+      dataset %>%
+        group_by(!!!by_vars) %>%
+        filter(!!enexpr(x)) %>%
+        mutate(!!!y) %>%
```
Comment (bundfussr) [inline R/derive_extreme_event.R:125]: Could we add a temporary event number variable and add it to the `order` in the `filter_extreme()` call? My intention is that the events are handled like the conditions in a `case_when()`. I.e., the first event from the `events` list is selected and the `order` and `mode` arguments are only used for the case that there is more than one observation of a particular event. Then the user would not need to create an order variable for the events.
Consider for example that `AVAL` should be `0` for the `"Missing"` event. Currently `AVAL` could not be used for `order` but the user would need to create a temporary order variable and remove it after the `derive_extreme_event()` call.
Does this make sense?

File: R/derive_extreme_event.R [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  assert_order_vars(order)
+  mode <- assert_character_scalar(mode, values = c("first", "last"), case_sensitive = FALSE)
+  check_type <-
+    assert_character_scalar(
+      check_type,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  assert_varval_list(set_values_to)
+
```
[the remark this answers is not in the document]
Comment (kaz462) [reply R/derive_extreme_event.R:114]: thanks for catching this! I copied those from  `derive_extreme_records()`, should I add `dataset` assertion for `derive_extreme_records()` as well?

File: R/derive_extreme_event.R (hunk 1 above)
Comment (kaz462) [reply R/derive_extreme_event.R]: I updated the documentation as "Add the first available record from `events`...", so that 
1. for all selected events, always choose the first available event by entering order;
2. `order`/`mode` in this function is for filtering extreme record within a certain event (when there is more than one observation of a particular event)

File: R/derive_extreme_event.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_extreme_event.R:120]: Yes, please add.

File: R/derive_extreme_event.R [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  set_values_to_ls <- map(events, "set_values_to")
+  event_order <- map(seq_len(length(events)), function(x) x)
+
+  selected_records_ls <- purrr::pmap(
+    list(condition_ls, set_values_to_ls, event_order),
+    function(x, y, z) {
+      dataset %>%
+        group_by(!!!by_vars) %>%
+        filter(!!enexpr(x)) %>%
+        mutate(!!!y, event_no = z)
```
Comment (bundfussr) [inline R/derive_extreme_event.R:133]: Could you use `get_new_tmp_var()` to create a name for the temporary variable? Then it can not happen that an existing variable is accidentally overwritten.

File: R/derive_extreme_event.R [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    ungroup()
+
+  ## Filter extreme among all events
+  new_obs <- selected_records_extreme %>%
+    filter_extreme(
+      by_vars = by_vars,
+      order = exprs(event_no),
+      mode = "first",
+      check_type = check_type
+    ) %>%
```
Comment (bundfussr) [inline R/derive_extreme_event.R:164]: Do we need to split the selection? I wonder if we could do it with one call. E.g.,
```
if (mode == "first") {
  event_no_expr <- expr(event_no)
} else {
  event_no_expr <- expr(desc(event_no))
}
new_obs <- selected_records_extreme %>%
    filter_extreme(
      by_vars = by_vars,
      order = exprs_c(event_no_expr, order),
      mode = mode,
      check_type = check_type
    ) 
```

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_extreme_event.R (hunk 4 above)
Comment (kaz462) [reply R/derive_extreme_event.R:164]: should `mode` be applied to `order` instead of `event_no`? in the above example, `mode=first` or `last` will return the same results, e.g., `first` - order by event 1,2,3 and choose the first one (event_no=1); `last` - order by event 3,2,1 and choose the last one (event_no=1).
if we combine the two calls, maybe we can fix `mode="first"` in `filter_extreme`, but I don't know how to "reverse" the `order` (like above example for `event_no`) when it contains multiple variables

Comment (kaz462) [reply R/derive_extreme_event.R]: I incorporated the `order` info to a temporary obs number `tmp_obs`  (by `by_vars` and `tmp_event_no`) as a workaround

File: R/derive_extreme_event.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaz462) [reply R/derive_extreme_event.R:129]: added for `derive_extreme_records()` in the same PR

File: R/derive_extreme_event.R (hunk 4 above)
Comment (bundfussr) [reply R/derive_extreme_event.R]: Your approach works as well.

**Review by bundfussr — APPROVED** [review]
